### PR TITLE
tstime/mono: add new package, use everywhere

### PR DESCRIPTION
**Pull request by josharian**
We call time.Now in 4-6 places per packet.
At 50+ns per call, it adds up enough to show up in pprof.

time.Now is twice as expensive as we need,
because it gets both wall time and monotonic time.
We need only monotonic time.

This PR introduces tstime/mono.Time and
converts all but one per-packet time.Now call
to use it. (The "but one" is in wireguard-go.)

tstime.Mono is about 50% faster than time.Now,
at the cost of using a go:linkname to grab
the runtime's internal monotonic clock routine.

This shrinks the "current time" pprof box,
but doesn't eliminate it.

The next step here is to decide which of our
uses of time would accept a granularity of 1-4ms,
and switch them to use a coarse monotonic clock.
That's faster yet than the regular monotonic clock,
but it also means making our own copies of
the package runtime assembly that gets the time.

### CODE REVIEW

**Review by crawshaw — COMMENTED** [review]
General strategy looks good. It's somewhat unfortunate how big a diff is created by `s/time.Time/tstime.Mono/`, but I think it's for the best that we don't do something tricky like a rename import to keep using time.Time.

A quick survey of these uses of time.Now suggest they could all be replaced with a coarse variant.

**Comment by bradfitz** [comment]
Bikeshed: package `tstime/mono` (or even something outside of `tailscale.com` entirely) and then `mono.Time` as the type? Same length as `time.Time`.

**Comment by josharian** [comment]
> Bikeshed: package tstime/mono (or even something outside of tailscale.com entirely) and then mono.Time as the type? Same length as time.Time.

Done.

> A quick survey of these uses of time.Now suggest they could all be replaced with a coarse variant.

I recall seeing one (DERP timing?) that appeared to want millisecond resolution. Maybe that'd be OK with coarse time? Not sure. I'll look again.

Once this is in (after the 1.12 release!), I'll pull in my coarse time implementations and do a bit more retrofitting. I plan to use the same type (mono.Time) for coarse time, with the only difference being calling `mono.CoarseNow` and `mono.CoarseSince` instead of `mono.Now` and `mono.Since`, so the diff should be much smaller.


Now to figure out why tests pass locally but not on CI...

**Comment by josharian** [comment]
Ready for review at last.

File: ipn/ipnstate/ipnstate.go [hunk 1]
```diff
@@ -90,7 +91,7 @@ type PeerStatus struct {
 	RxBytes       int64
 	TxBytes       int64
 	Created       time.Time // time registered with tailcontrol
-	LastWrite     time.Time // time last packet sent
+	LastWrite     mono.Time // time last packet sent
```
Comment (bradfitz) [inline ipn/ipnstate/ipnstate.go:94]: How does this serialize in JSON in `tailscale status --json` ?

Comment (josharian) [reply ipn/ipnstate/ipnstate.go:94]: Badly. I've changed it to format as a time.Time, for backwards compatibility. Of course, if the wall clock changes, it won't be responsive to that change.

File: net/tstun/wrap.go [hunk 2]
```diff
@@ -64,7 +65,7 @@ type Wrapper struct {
 
 	closeOnce sync.Once
 
-	lastActivityAtomic int64 // unix seconds of last send or receive
+	lastActivity mono.Time // time of last send or receive, accessed atomically
```
Comment (bradfitz) [inline net/tstun/wrap.go:68]: keep the old variable name? the Atomic suffix is ugly and scary which is a feature.

I guess you didn't like the stutter in `lastActivityAtomic.LoadAtomic` etc? I'm fine with it if you are.

File: tstime/mono/mono.go [hunk 3]
```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package mono provides fast monotonic time.
```
Comment (bradfitz) [inline tstime/mono/mono.go:5]: Maybe flesh out this package comment a bit with some numbers and a big scary warning that it should almost never be used and std time is plenty fast?

**Review by bradfitz — APPROVED** [review]

File: tstime/rate/rate.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+// bursts of at most b tokens.
+func NewLimiter(r Limit, b int) *Limiter {
+	if b < 1 {
+		panic("bad burst, must be at least 1")
+	}
+	return &Limiter{limit: r, burst: float64(b)}
+}
+
+// AllowN reports whether an event may happen now.
+func (lim *Limiter) Allow() bool {
```
Comment (crawshaw) [inline tstime/rate/rate.go:59]: If you plan to add a `mono.NowCoarse`, then I think it's worth making this Allow take a `now` parameter, so the caller can choose if they want coarse or fine-grained timing on their token bucket

**Review by crawshaw — APPROVED** [review]

File: net/tstun/wrap.go (hunk 2 above)
Comment (josharian) [reply net/tstun/wrap.go:68]: Yeah, I was removing the stutter, but don't feel strongly. Re-added.

File: tstime/mono/mono.go (hunk 3 above)
Comment (josharian) [reply tstime/mono/mono.go:5]: Done

File: tstime/rate/rate.go (hunk 4 above)
Comment (josharian) [reply tstime/rate/rate.go:59]: Ack, although I'd rather do that as part of adding coarse mono time.

**Pull request merged**